File: 3150-275-916-word-subsets/3150-275-916-word-subsets.js
var wordSubsets = function(words1, words2) {
    const solve = (s) => {
        const freq = Array(26).fill(0);
        for (const ch of s) {
            freq[ch.charCodeAt(0) - 'a'.charCodeAt(0)]++;
        }
        return freq;
    };

    const frq = Array(26).fill(0);
    for (const word of words2) {
        const tmp = solve(word);
        for (let i = 0; i < 26; i++) {
            frq[i] = Math.max(frq[i], tmp[i]);
        }
    }

    const result = [];
    for (const word of words1) {
        const tmp = solve(word);
        if (frq.every((count, i) => count <= tmp[i])) {
            result.push(word);
        }
    }

    return result;
};

File: 3150-275-916-word-subsets/3150-275-916-word-subsets.java
import java.util.*;

class Solution {
    private int[] solve(String s) {
        int[] freq = new int[26];
        for (char ch : s.toCharArray()) {
            freq[ch - 'a']++;
        }
        return freq;
    }

    public List<String> wordSubsets(String[] words1, String[] words2) {
        int[] frq = new int[26];
        for (String word : words2) {
            int[] tmp = solve(word);
            for (int i = 0; i < 26; i++) {
                frq[i] = Math.max(frq[i], tmp[i]);
            }
        }

        List<String> result = new ArrayList<>();
        for (String word : words1) {
            int[] tmp = solve(word);
            boolean isUniversal = true;
            for (int i = 0; i < 26; i++) {
                if (frq[i] > tmp[i]) {
                    isUniversal = false;
                    break;
                }
            }
            if (isUniversal) {
                result.add(word);
            }
        }

        return result;
    }
}
